design/cluster-api-provider-metal3/multi-tenancy_contract.md

```diff
@@ -0,0 +1,300 @@
+
+#### Story 3
+
+As a user of CAPM3, I want to make use of the BareMetalHosts that another team
+has created and made available to me, without necessarily sharing the same
+cluster.
+
+#### Story 4
+
+As a one-man-show, I want to continue using the current single-tenant setup with
+in-cluster credentials for accessing the BareMetalHosts. This is especially
+important for bootstrapping a self-hosted management cluster.
+
+#### Story 5
+
+As an advanced user of CAPM3, I want to be able to build clusters spanning over
+multiple sets of BareMetalHosts, each with their own credentials, for improved
+resilience.
+
+### Implementation Details
+
+The new Metal3Cluster field will be `.spec.identityRef` with sub-fields `name`
+and `context` (optional). The `name` should point to a secret in the same
+namespace as the Metal3Cluster, containing a kubeconfig file. The optional
+`context` should point to a context in the kubeconfig file. If the `context` is
+not specified, the current context will be used.
+
+For the Metal3MachineTemplate, the new field will be
+`.spec.template.spec.identityRef` with the same sub-fields as the Metal3Cluster.
+This will propagate to the Metal3Machine's `.spec.identityRef`.
+
+If the `identityRef` is not specified, CAPM3 should use in-cluster credentials
+for accessing the BareMetalHosts (current behavior). If the `identityRef` is
+specified on the Metal3Cluster, but not on the Metal3MachineTemplate, CAPM3
+should use the credentials specified in the Metal3Cluster for all
+Metal3Machines.
+
+CAPM3 should make use of the kubeconfig file to access the BareMetalHosts
+related to the Metal3Cluster. The kubeconfig file can point to a namespace in
+the same cluster or to a remote cluster. Secrets for metadata, networkdata and
+userdata must also be created next to the BareMetalHosts.
+
+Example:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3Cluster
+metadata:
+  name: cluster-1
+spec:
+  identityRef:
+    name: cluster-1-kubeconfig
+    context: cluster-1-context
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: cluster-1-controlplane-template
+spec:
+  template:
+    spec:
+      identityRef:
+        name: control-plane-bmh-kubeconfig
+        context: default
+```
+
+The status of the Metal3Machine should be exteded with the information from the
+BareMetalHost. This is similar to how other providers work, as they propagate
+the relevant information from cloud resources back the the InfraMachine. For the
+Metal3Machine, we should add the following fields from the BareMetalHost:
+
+- `.status.provisioning` on the BMH would become
+  `.status.bareMetalHost.provisioning`
+- `.status.poweredOn` on the BMH would become `.status.bareMetalHost.poweredOn`
+- `.status.errorCount` on the BMH would become
+  `.status.bareMetalHost.errorCount`
+- `.status.errorMessage` on the BMH would become
+  `.status.bareMetalHost.errorMessage`
```

Comment:
Something to think about: we might expose internal details through error messages.

Reply:
Do you see a risk that this could be sensitive or just something the user might learn to rely on even though they should not? Or that it would be irrelevant to the user?

I'm thinking that we will need to expose some kind of information in any case and this seems to be similar to what other providers do. In my experience it is seldom something the user can "do" anything with, but it can often be useful to determine if there is a user error or rather some issue with the infrastructure.

Reply:
> Do you see a risk that this could be sensitive or just something the user might learn to rely on even though they should not? Or that it would be irrelevant to the user?

Yes :)

I don't have an easy solution to this. Errors from Ironic can be vital, e.g. to understand that RootDeviceHints are wrong. Maybe it's something we should simply document.